python/lsst/daf/ingest/ingestCatalogTask.py

```diff
@@ -0,0 +1,580 @@
+    """
+
+    def __init__(self, sqlTypeCallable, formatValueCallable):
+        """Store the field formatting information.
+        """
+        self.sqlTypeCallable = sqlTypeCallable
+        self.formatValueCallable = formatValueCallable
+
+    def sqlType(self, field):
+        """Return the SQL type for the given field's values.
+        """
+        return self.sqlTypeCallable(field)
+
+    def formatValue(self, value):
+        """Return a string representation of value suitable for use in an
+        INSERT/REPLACE statement (not a CSV file).  Nones are always
+        transformed to SQL NULLs.
+        """
+        if value is None:
+            return "NULL"
+        return self.formatValueCallable(value)
+
+
+def _formatNumber(formatString, number):
+    """Helper function for formatting floating point numbers. NaNs and
+    infinities are converted to NULL, because MySQL does not seem to support
+    NaN/Inf in FLOAT or DOUBLE columns.
+    """
+    if math.isnan(number) or math.isinf(number):
+        return "NULL"
+    return formatString.format(number)
+
+def _formatString(string):
+    """Helper function for formatting a string. Embedded backslashes and single
+    quotes are backslash-escaped.
+    """
+    return "'" + string.replace("\\", "\\\\").replace("'", "\\'") + "'"
+
+def _formatArray(formatChar, array):
+    """Auxiliary function for formatting an array. The array elements are packed
+    into a sequence of bytes, with bytes comprising individual elements arranged
+    in little endian order. This sequence is then transformed into a MySQL
+    hexadecimal literal and returned.
+    """
+    bytestring = struct.pack("<" + str(len(array)) + formatChar, *array)
+    return "x'" + bytestring.encode('hex_codec') + "'"
+
+
+def _sqlTypeForString(field):
+    """Helper function for computing the SQL column type of a string valued field.
+    """
+    sz = field.getSize()
+    if sz > 65535:
+        # If necessary, longer strings could be ingested as TEXT.
+        raise RuntimeError("String field is too large for ingestion")
+    elif sz == 0:
+        raise RuntimeError("String field has zero size")
+    # A string containing trailing spaces cannot round-trip to a CHAR
+    # column and back. Therefore, use VARCHAR. Also, assume strings are
+    # ASCII for now.
+    return "VARCHAR({}) CHARACTER SET ascii COLLATE ascii_bin NOT NULL".format(sz)
+
+def _sqlTypeForArray(formatChar, field):
+    """Helper function for computing the SQL column type of an array valued field.
+    See https://docs.python.org/2/library/struct.html#format-characters for possible
+    format characters. The formatChar passed in must correspond to the array element
+    type.
+    """
+    sz = field.getSize()
+    if sz == 0:
+        return "BLOB NOT NULL"
+    sz *= struct.calcsize("<" + formatChar)
+    if sz > 65535:
+        raise RuntimeError("Array field is too large for ingestion")
+    return "BINARY({}) NOT NULL".format(sz)
```

Comment:
Because `VARBINARY` has to store a length for every value and `BINARY` does not. And, unlike `String` fields, `Array` fields created with a non-zero size really are fixed size. `CHAR(n)` doesn't work for `String` fields because MySQL pads with trailing spaces to reach the specified storage size (`n`), and removes trailing spaces on retrieval. So roundtrips with `String` values must use `VARCHAR` for correctness. Another subtlety is that I have to use `BLOB` rather than `VARBINARY(large number)` for variable length `Array` fields. That's because the maximum size passed to `VARBINARY` seems to count against ones row size, which means that MySQL will complain about exceeding the 64k row size limit for schemas containing just a couple `VARBINARY` columns with large maximum sizes. `BLOB`s store their data off to the side, and don't have the same problem.